### ISSUE TITLE
dynamic shapes support for __torch_dispatch__ subclasses + torch.compile

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#114374]
Dynamic shapes work with subclasses in "basic" cases (where the subclass and its inner tensor(s) have the exact same shape), but blow up in more complicated cases.

Design doc on the plan is [here](https://docs.google.com/document/d/1ipSxcTzEMMOAPvxP-YJlD5JBZZmIGgh8Q34ixtOUCRo/edit#heading=h.3px8g3br0skz), authored with @ezyang and @jbschlosser.

We should make sure that several "interesting" subclass examples work:
- NestedTensor (ragged dim is dynamic)
- Float8Tensor (scales tensor is static)
- SparseTensor (nnz dim of the values tensor is always dynamic)

**Comment by bdhirsh** [comment]
Another "suboptimal" aspect of dynamic shapes + subclasses today is that `_make_wrapper_subclass` has two constructors, where:

(1) they seem redundant (one takes in kwargs, the other takes positional args, but otherwise they are very similar)

(2) the kwargs one does not worth with dynamic shapes, but the args one does.

Fixing this is not critical to getting dynamic shapes working with subclasses, but it will fix a footgun and improve subclass UX

**Comment by bdhirsh** [comment]
From talking to @wanchaol, it sounds like for DTensor, we might have a situation where the outer DTensor's **strides** are independent of the values of the sizes/strides of the inner tensors. So we might need to eventually pass in `outer_strides` to `__tensor_unflatten__`, too

**Comment by jon-chuang** [comment]
> the outer DTensor's strides are independent of the values of the sizes/strides of the inner tensors.

Hmm, what do these outer strides represent?

**Comment by jbschlosser** [comment]
Another issue: we should properly install TENSOR_MATCH guards on subclass inner tensors. I started doing this but ran into a problem with `NestedTensor` purposefully aliasing `offsets` across subclass instances. This triggers:

https://github.com/pytorch/pytorch/blob/4d4093a5de830f2382349be5d91b250e6fec93b5/torch/csrc/dynamo/guards.cpp#L402-L408

**Comment by jbschlosser** [comment]
Another issue: I tried introducing an `outer_storage_offset` as well, but storage_offset handling for subclasses is pretty broken in PT2. Note that we're ignoring `storage_offset` here in this helper function used for meta-ifying inner tensors:

https://github.com/pytorch/pytorch/blob/597d3fb86a2f3b8d6d8ee067e769624dcca31cdb/torch/_subclasses/meta_utils.py#L455-L466

**Comment by jon-chuang** [comment]
> we should properly install TENSOR_MATCH guards on subclass inner tensors

Why is that? Is the nested tensor outer meta match not enough to fully determine the compilation pathway?

**Comment by jbschlosser** [comment]
> Why is that? Is the nested tensor outer meta match not enough to fully determine the compilation pathway?

That's right. This will be the case when the inner tensor shapes are not fully implicated by the outer tensor shape. In practice, this usually isn't the case, but I think a sparse tensor would be a good example here.

Say you have a sparse outer tensor of some constant shape with inner tensors of `values` / `indices` with shapes determined by the number of nonzero (nnz) entries. If there's any graph where we need to specialize on `nnz` (maybe if the nonzero entries are iterated over), then we'd want to recompile if this changes.
